src/object.c

```diff
@@ -269,53 +308,78 @@ long long objectGetExpire(const robj *val) {
 
 /* This functions may reallocate the value. The new allocation is returned and
  * the old object's reference counter is decremented and possibly freed. Use the
- * returned object instead of 'val' after calling this function. */
-robj *objectSetExpire(robj *val, long long expire) {
-    if (val->hasexpire) {
+ * returned object instead of 'o' after calling this function. */
+robj *objectSetExpire(robj *o, long long expire) {
+    if (o->hasexpire) {
         /* Update existing expire field. */
-        unsigned char *data = (void *)(val + 1);
+        unsigned char *data = objectEmbeddedData(o);
         *(long long *)data = expire;
-        return val;
+        return o;
     } else if (expire == -1) {
```

Comment:
If we are changing stuff...
```suggestion
    } else if (expire == EXPIRY_NONE) {
```